ValveResourceFormat/IO/FileExtract.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.IO;
+using System.Text;
+using SkiaSharp;
+using ValveResourceFormat;
+using ValveResourceFormat.ResourceTypes;
+
+namespace ValveResourceFormat.IO
+{
+    public static class FileExtract
+    {
+        public static Span<byte> Extract(Resource resource)
+        {
+            Span<byte> data;
+
+            switch (resource.ResourceType)
+            {
+                case ResourceType.Panorama:
+                case ResourceType.PanoramaLayout:
+                case ResourceType.PanoramaScript:
+                case ResourceType.PanoramaStyle:
+                    data = ((Panorama)resource.Blocks[BlockType.DATA]).Data;
+                    break;
+
+                case ResourceType.Sound:
+                    data = ((Sound)resource.Blocks[BlockType.DATA]).GetSound();
+
+                    break;
+
+                case ResourceType.Texture:
+                    var bitmap = ((Texture)resource.Blocks[BlockType.DATA]).GenerateBitmap();
+                    var image = SKImage.FromBitmap(bitmap);
+
+                    using (var ms = new MemoryStream())
+                    {
+                        using (var imageData = image.Encode(SKEncodedImageFormat.Png, 100))
+                        {
+                            imageData.SaveTo(ms);
+                        }
+
+                        data = ms.ToArray();
+                    }
+
+                    break;
+
+                case ResourceType.Particle:
+                case ResourceType.Mesh:
+                    // Wrap it around a KV3File object to get the header.
+                    data = Encoding.UTF8.GetBytes(((BinaryKV3)resource.Blocks[BlockType.DATA]).GetKV3File().ToString());
+                    break;
+
+                // These all just use ToString() and WriteText() to do the job
+                case ResourceType.SoundEventScript:
+                    data = Encoding.UTF8.GetBytes(resource.Blocks[BlockType.DATA].ToString());
+                    break;
+
+                default:
+                    Console.WriteLine("-- (I don't know how to dump this resource type)"); // TODO: What do we do with this
+                    data = Encoding.UTF8.GetBytes(resource.Blocks[BlockType.DATA].ToString());
```

Comment:
Switching from not dumping on decompiler to dumping potentially an error is a change in behaviour, is this what we want?

Reply:
That's what I mentioned in the initial post, different code paths handled this differently, and I'm not sure how we want to handle it here.

Reply:
When decompiling from vpk here: https://github.com/SteamDatabase/ValveResourceFormat/blob/41f467ca61cc0fe78b10f01522bbaef25ba71703/Decompiler/Decompiler.cs#L660

it will fallback to dumping the file as-is when an exception is thrown. Other code paths would error out (if DATA block is missing, or it has no ToString), so we should be fine?